lumeer-core/src/main/kotlin/io/lumeer/core/task/executor/operation/stage/RemoveDocumentsStage.kt

```diff
@@ -0,0 +1,149 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package io.lumeer.core.task.executor.operation.stage
+
+import io.lumeer.api.model.*
+import io.lumeer.core.adapter.CollectionAdapter
+import io.lumeer.core.adapter.DocumentAdapter
+import io.lumeer.core.adapter.LinkInstanceAdapter
+import io.lumeer.core.adapter.LinkTypeAdapter
+import io.lumeer.core.facade.FunctionFacade
+import io.lumeer.core.task.RuleTask
+import io.lumeer.core.task.executor.ChangesTracker
+import io.lumeer.core.task.executor.operation.DocumentRemovalOperation
+import io.lumeer.core.task.executor.operation.OperationExecutor
+import io.lumeer.core.util.Utils
+import io.lumeer.engine.api.event.RemoveDocument
+import io.lumeer.engine.api.event.RemoveLinkInstance
+import java.util.function.Function
+import java.util.stream.Collectors.toMap
+
+class RemoveDocumentsStage(executor: OperationExecutor) : Stage(executor) {
+
+    private val collectionAdapter = CollectionAdapter.createFromDaoSnanpshot(task.daoContextSnapshot)
+    private val linkTypeAdapter = LinkTypeAdapter.createFromDaoSnapshot(task.daoContextSnapshot)
+    private val documentAdapter = DocumentAdapter.createFromDaoSnapshot(task.daoContextSnapshot)
+    private val linkInstanceAdapter = LinkInstanceAdapter.createFromDaoSnapshot(task.daoContextSnapshot)
+    private val functionFacade: FunctionFacade = task.functionFacade
+    private val taskProcessingFacade = task.getTaskProcessingFacade(taskExecutor, functionFacade)
+    private val purposeChangeProcessor = task.purposeChangeProcessor
+
+    override fun call(): ChangesTracker {
+        // remove documents
+        processRemoveDocuments(operations.filter { operation -> operation is DocumentRemovalOperation }.map { operation -> operation as DocumentRemovalOperation })
+
+        return changesTracker
+    }
+
+    private fun processRemoveDocuments(operations: List<DocumentRemovalOperation>) {
+        if (operations.isEmpty()) {
+            return
+        }
+
+        // collect all information that we will need during the process
+        val documents = operations.map { it.entity }
+        val updatedCollectionIds = documents.map { it.collectionId }
+        val updatedCollections = task.daoContextSnapshot.collectionDao.getCollectionsByIds(updatedCollectionIds)
+
+        val removedLinks = task.daoContextSnapshot.linkInstanceDao.getLinkInstancesByDocumentIds(documents.map { it.id }.toSet())
+        val updatedLinkTypeIds = removedLinks.map { it.linkTypeId }
+        val updatedLinkTypes = task.daoContextSnapshot.linkTypeDao.getLinkTypesByIds(updatedLinkTypeIds.toSet())
+
+        // remove all data - documents, links, favorites, comments
+        removeLinks(removedLinks, documents, updatedLinkTypes.stream().collect(toMap({ it.id }, Function.identity())))
+        removeDocuments(documents, updatedCollections.stream().collect(toMap({ it.id }, Function.identity())))
```

Comment:
`.stream().collect` etc is not needed in kotlin, you can directly use `map { it.id }` and thats it